File: checker2/Rover/src/aCheck/Violation.java
package aCheck;

import model.Group;
import model.Interaction;
import checkers.Property;
import java.util.ArrayList;

public class Violation{ 

    private Property prop; //the property that is violated
    private ArrayList<Group> groupsViolating; // if this property is violated by a group, this stores the list of groups violating this property
    private boolean tiesToInteraction; //properties are either related to groups or to an interaction, this is saying true if the prop is an interaction property
    private String type; // the class of violation it falls under
    private String description; //description of this violation

    public Violation(Property prop){
        this.prop = prop;//the property contains the type, category and description
        groupsViolating = new ArrayList<Group>();
        type = prop.getCategory();
        if(prop.getTies().equals("interaction")){
            tiesToInteraction = true;
        }else{
            tiesToInteraction = false;
        }
    }

    public Violation(String type, boolean tiesToInteraction, String description){  //a special violation indexed by the type and has no property
        this.prop = null;//since we dont tie this to a property, we have to add type and description
        groupsViolating = new ArrayList<Group>();
        this.type = type;
        this.description = description;
        if(tiesToInteraction){
            this.tiesToInteraction = true;
        }else{
            this.tiesToInteraction = false;
        }
    }

    public String getDescription(){
        return description;
    }

    public String getType(){
        return type;
    }

    public String getPropType(){
        return prop.toString();
    }

    public String propDesc(){
        return prop.getDescription();
    }

    public String toString(){
        String returnString = "";
        returnString += "TYPE: " + type + "\n" + " \nDesc: " + description + "\n";
        if(!tiesToInteraction){
            for(Group g : groupsViolating){
                   returnString += "\nGROUP: " + g.toString() ;
            }
         }
        return returnString;


    }

    public void addGroupViolating(Group g){
        if(tiesToInteraction){
            //System.out.println("WARNING: tried to add a group to an interaction only property, see Violation class");
            return;
        }
        groupsViolating.add(g);
    }

    public boolean isViolatingAtInteractionLevel(){
        return tiesToInteraction;
    }

    public boolean isViolatingAtGroupLevel(){
        return !tiesToInteraction;
    }



    public Property getProperty(){
        return prop;
    }

    public ArrayList<Group> getGroupsViolating(){
        return groupsViolating;
    }




}

File: client/src/app/models/violation.ts
export class Violation {
    category: string;   //-either "group" or "interaction"  --whether this is a group level violation (has associated group) or an interaction level violation (is violated somewhere in the interaction but exact location cant be pinpointed)
    type: string   // -- the type of property being violated
    description: string //--the description of the property being violated
    violatorGroups: string[] = []; //  --[groupName]   list of name of the group (or groups) violating this property, will be empty if category is violation

    constructor(category: string, type: string, description: string) {
        this.category = category;
        this.type = type;
        this.description = description;
        this.violatorGroups = [];
    }

    addGroupViolating(groupName: string) {
        this.violatorGroups.push(groupName);
    }

    setGroupsViolating(groupIDs: string[]) {
        this.violatorGroups = groupIDs;
    }

    getGroupsViolating() {
        return this.violatorGroups;
    }
}